cjdns-bytes/src/utils.rs

```diff
@@ -1,150 +1,224 @@
+        fn test_reader() {
+            let bytes = vec![0u8; 64];
+            let mut reader = Reader::new(&bytes);
+
+            let array_32 = reader.read_array_32().expect("bytes are read out");
+            assert_eq!(array_32, [0; 32]);
+            let num64 = reader.read_u64_be().expect("bytes are read out");
+            assert_eq!(num64.to_be_bytes(), [0; 8]);
+            let num32 = reader.read_u32_be().expect("bytes are read out");
+            assert_eq!(num32.to_be_bytes(), [0; 4]);
+            let num16 = reader.read_u16_be().expect("bytes are read out");
+            assert_eq!(num16.to_be_bytes(), [0; 2]);
+            let num8 = reader.read_u8().expect("bytes are read out");
+            assert_eq!(num8, 0);
+            // checking read out of bounds
+            assert!(reader.read_slice(20).is_err());
+            // reading last bytes
+            reader.read_slice(17).expect("bytes are read out");
+            assert_eq!(reader.len(), 0);
+        }
+
+        #[test]
+        fn test_read_all() {
+            let bytes = vec![0; 32];
+            let mut reader = Reader::new(&bytes);
+
+            assert_eq!(bytes.as_slice(), reader.pick_remainder());
+            let _ = reader.read_u16_be();
+            assert_eq!(&bytes[2..], reader.pick_remainder());
+            let _ = reader.read_u16_be();
+            assert_eq!(&bytes[4..], reader.read_remainder());
+            assert_eq!(0, reader.len())
+        }
+
+        #[test]
+        #[should_panic]
+        fn test_custom_read() {
+            let bytes = vec![0; 32];
+            let mut reader = Reader::new(&bytes);
+
+            assert!(reader.read(SizePredicate::Exact(35), |_| Ok(())).is_err());
+            assert!(reader.read(SizePredicate::NotLessThan(35), |_| Ok(())).is_err());
+            // panics: trying to read more than stated
+            let _ = reader.read(SizePredicate::Exact(32), |r| {
+                let invalid_res = r.read_slice(35)?;
+                Ok(invalid_res)
+            });
+            // panics: trying to read less than stated
+            let _ = reader.read(SizePredicate::Exact(32), |r| {
+                let invalid_res = r.read_slice(20)?;
+                Ok(invalid_res)
+            });
+            // panics: trying to read less than stated
+            let _ = reader.read(SizePredicate::NotLessThan(32), |r| {
+                let invalid_res = r.read_slice(20)?;
+                Ok(invalid_res)
+            });
+            // panics: trying to read more than stated
+            let _ = reader.read(SizePredicate::NotLessThan(32), |r| {
+                let invalid_res = r.read_slice(35)?;
+                Ok(invalid_res)
+            });
+
+        }
     }
 }
 
-impl Writer {
-    pub fn with_capacity(capacity: usize) -> Self {
-        Self(Vec::with_capacity(capacity))
-    }
+mod writer {
+    pub struct Writer(Vec<u8>);
 
-    pub fn into_vec(self) -> Vec<u8> {
-        self.0
-    }
+    impl Writer {
```

Comment:
Actually it might me useful to implement the same closure-driven `write` method as above with the reader.
So you call it like `writer.write(MyHeader::SIZE, |w| { w.write_u8(...)... })` so it will check that there was exactly N bytes written. Just the same idea, same sanity check.

Reply:
discussed